cns/restserver/ipam.go

```diff
@@ -92,40 +92,58 @@ func (service *HTTPRestService) releaseIPConfigHandler(w http.ResponseWriter, r
 	return
 }
 
-// MarkIPAsPendingRelease will mark IPs to pending release state.
-func (service *HTTPRestService) MarkIPAsPendingRelease(numberToMark int) (map[string]cns.IPConfigurationStatus, error) {
-	allReleasedIPs := make(map[string]cns.IPConfigurationStatus)
-	// Ensure PendingProgramming IPs will be release before Available ones.
-	ipStateTypes := [2]string{cns.PendingProgramming, cns.Available}
-
+// MarkIPAsPendingRelease will set the IPs which are in PendingProgramming or Available to PendingRelease state
+// It will try to update [totalIpsToRelease]  number of ips.
+func (service *HTTPRestService) MarkIPAsPendingRelease(totalIpsToRelease int) (map[string]cns.IPConfigurationStatus, error) {
+	pendingReleasedIps := make(map[string]cns.IPConfigurationStatus)
 	service.Lock()
 	defer service.Unlock()
-	for _, ipStateType := range ipStateTypes {
-		pendingReleaseIPs := service.markSpecificIPTypeAsPending(numberToMark, ipStateType)
-		for uuid, pependingReleaseIP := range pendingReleaseIPs {
-			allReleasedIPs[uuid] = pependingReleaseIP
-		}
-		numberToMark -= len(pendingReleaseIPs)
-		if numberToMark == 0 {
-			return allReleasedIPs, nil
-		}
+
+	var count int
+	for uuid, ipcConfig := range service.PodIPConfigState {
+        if ipcConfig.State == cns.PendingProgramming {
+            if err := service.updateIPConfigState(uuid, cns.PendingRelease); err != nil {
+                return nil, err
+            }
+
+            pendingReleasedIps[uuid] = ipcConfig
+			count++
```

Comment:
Yes, I checked this link and thought it may be more perfomant to use the count, [https://stackoverflow.com/questions/26634554/go-multiple-len-calls-vs-performance](https://stackoverflow.com/questions/26634554/go-multiple-len-calls-vs-performance). (read the last part) where someone also tried local slice perf

---
In reply to: [584960396](https://github.com/Azure/azure-container-networking/pull/806#discussion_r584960396) [](ancestors = 584960396)